airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py

```diff
@@ -0,0 +1,451 @@
+                self.client.schema.delete_class(class_name=class_name)
+
+            if current_class is None or existing == "replace":
+                self.client.schema.create_class(class_object)
+                self.logger.info(f"Created/updated class {class_name}")
+
+    def prepare_data_for_ingestion(
+        self, dfs: list[pd.DataFrame], class_name: str, existing: str, uuid_column: str, vector_column: str
+    ) -> tuple[pd.DataFrame, str]:
+        """
+        Prepares data for ingestion into Weaviate.
+
+        :param dfs: A list of dataframes from downstream dynamic tasks.
+        :param class_name: The name of the class to import data.
+        :param existing: Strategy to handle existing data ('skip', 'replace', 'upsert').
+        :param uuid_column: Name of the column containing the UUID.
+        :param vector_column: Name of the column containing the vector data.
+        :return: A concatenated and processed DataFrame ready for ingestion.
+        """
+        if existing not in ["skip", "replace", "upsert"]:
+            raise AirflowException("Invalid parameter for 'existing'. Choices are 'skip', 'replace', 'upsert'")
+
+        df = pd.concat(dfs, ignore_index=True)
+
+        if uuid_column is None:
+            column_names = df.columns.to_list()
+            column_names.sort()
+            df = df[column_names]
+
+            self.logger.info("No uuid_column provided. Generating UUIDs for ingest.")
+            if "id" in column_names:
+                raise AirflowException("Property 'id' already in dataset. Consider renaming or specify 'uuid_column'.")
+            else:
+                uuid_column = "id"
+
+            df[uuid_column] = df.drop(columns=[vector_column], inplace=False, errors="ignore").apply(
+                lambda row: generate_uuid5(identifier=row.to_dict(), namespace=class_name), axis=1
+            )
+
+            df.drop_duplicates(inplace=True)
+
+        return df, uuid_column
+
+    def handle_upsert(self, df: pd.DataFrame, class_name: str, doc_key: str, uuid_column: str) -> (pd.DataFrame, dict):
+        """
+        Handles the 'upsert' operation for data ingestion.
+
+        :param df: The DataFrame containing the data to be upserted.
+        :param class_name: The name of the class to import data.
+        :param doc_key: The document key used for upsert operation.
+        :param uuid_column: The column name containing the UUID.
+        :return: The DataFrame filtered for objects to insert.
+        """
+        if doc_key is None:
+            raise AirflowException("Must specify 'doc_key' if 'existing=upsert'.")
+
+        if df[[doc_key, uuid_column]].duplicated().any():
+            raise AirflowException("Duplicate rows found. Remove duplicates before ingest.")
+
+        current_schema = self.client.schema.get(class_name=class_name)
+        doc_key_schema = [prop for prop in current_schema["properties"] if prop["name"] == doc_key]
+
+        if len(doc_key_schema) < 1:
```

Comment:
```suggestion
        if not doc_key_schema:
```

I guess this is for checking whether this list is empty?

Reply:
Done